File: pickr/lib/classes/games/briscola.dart
import 'package:pickr/classes/card.dart';
import 'package:pickr/classes/games/game.dart';
import 'package:pickr/enums/suits.dart';
import 'package:pickr/exceptions/out_of_bound_exception.dart';

import '../player.dart';

class Briscola extends Game {
  //
  /// Number of beginning cards for each player.
  final int _numcards = 3;

  /// Number of players.
  ///
  /// DEBUG: this value is set to 4 for debug purposes. Its real values should
  /// be selected by the user in the game options selection page.
  final int _numplayers = 4;

  /// Briscola suit selected for the game
  Suit _briscola;

  Briscola();

  @override
  void spread() {
    players.forEach((player) {
      player.hand = deck.picks(_numcards);
    });
  }

  @override
  void start() {
    //
    deck.init();

    deck.shuffle();

    _briscola = deck.cards.last.suit;

    spread();
  }

  @override
  GamingCard rule() {
    var winningCard = table.first;

    table.forEach((card) {
      if (card.val > winningCard.val) {
        // T T T
        if (winningCard.suit == _briscola && card.suit == _briscola) {
          winningCard = card;
        }
        // T F F
        if (winningCard.suit != _briscola && card.suit != _briscola) {
          winningCard = card;
        }
      }
      // F T F
      else if (card.suit == _briscola && winningCard.suit != _briscola) {
        winningCard = card;
      }
    });

    return winningCard;
  }

  @override
  void addPlayer({String id}) {
    if (players.length <= _numplayers - 1) {
      players.add(Player(id: id));
    } else {
      throw OutOfBoundException();
    }
  }

  @override
  void gameround() {
    players.forEach((player) {
      dropCard(card: player.drop());
    });

    var winningCard = rule();

    winner(winningCard);

    if (cards.length > _numplayers) {
      players.forEach((player) {
        player.pick(deck.pick());
      });
    }
  }
}

File: pickr-java/src/pickr/games/Briscola.java
package pickr.games;

import pickr.classes.GamingCard;
import pickr.classes.Player;
import pickr.enums.Suit;
import pickr.exceptions.OutOfBoundException;
import pickr.exceptions.WrongCardNumberException;

public class Briscola extends Game{
	
	private int numcards = 3;
	
	protected Suit briscola;
	
	public Briscola(int numplayers, int score) {
		super(numplayers, score);
	}
	

	@Override
	void spread() throws OutOfBoundException {
		for(Player player : players) {
			player.hand(deck.picks(numcards));
		}		
	}

	@Override
	public void start() throws WrongCardNumberException, OutOfBoundException { 	
		
		this.deck.init();
		this.deck.shuffle();
		
		briscola = deck.getCards().get(deck.getCards().size()-1).getSuit();
		
		this.spread();
		
		// this.deck.show();
	}


	@Override
	protected GamingCard rule() {
		
		GamingCard winningCard = this.getTable().get(0);
		
		for(GamingCard card : this.getTable()) {	
				
			// T T T
			if(card.getVal() > winningCard.getVal() && winningCard.getSuit() == briscola && card.getSuit() == briscola) {
				winningCard = card;
			}
			// T F F 
			if(card.getVal() > winningCard.getVal() && winningCard.getSuit() != briscola && card.getSuit() != briscola) {
				winningCard = card;
			}
			
			// F T F 
			 if(!(card.getVal() > winningCard.getVal()) && card.getSuit() == briscola && winningCard.getSuit() != briscola) {
				winningCard = card;
			}	
			
		}
		return winningCard;
	}
	



	public void round() throws OutOfBoundException {
		
		for(Player player : this.players) {
			this.dropCard(player.drop());
		}
		
		System.out.println("Briscola.round() : " + this.getTable());
		
		GamingCard winningCard = this.rule();
		
		this.winner(winningCard);	
		
		for(Player player : this.players) {
			player.pick(this.deck.pick());
		}		
	}
}
